### PR TITLE
ISSUE #978: decouple metaformat cmd

### PR DESCRIPTION
**Pull request by reddycharan**
Descriptions of the changes in this PR:

- introduce new commands - initnewcluster, nukeexistingcluster, whatisinstanceid
- add testcases initnewcluster and nukeexistingcluster at BookKeeperAdmin level

Master Issue: #978

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java [hunk 1]
```diff
@@ -421,94 +436,187 @@ public LayoutManager getLayoutManager(){
...
-            if (!availableNodeExists) {
-                zk.create(conf.getZkAvailableBookiesPath(), "".getBytes(Charsets.UTF_8),
-                    zkAcls, CreateMode.PERSISTENT);
+    public boolean prepareFormat() throws Exception {
+        boolean ledgerRootExists = null != zk.exists(conf.getZkLedgersRootPath(), false);
+        boolean availableNodeExists = null != zk.exists(conf.getZkAvailableBookiesPath(), false);
+        List<ACL> zkAcls = ZkUtils.getACLs(conf);
+        // Create ledgers root node if not exists
+        if (!ledgerRootExists) {
+            zk.create(conf.getZkLedgersRootPath(), "".getBytes(Charsets.UTF_8), zkAcls, CreateMode.PERSISTENT);
```
Comment (athanatos) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java]: Why not just call create() and ignore the EEXIST exception?  This variant has almost the same effect, but suffers from a race if the ledgerRoot is created between the check and here.  Same with available bookies node and the READONLY node.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/FlatLedgerManagerFactory.java [hunk 2]
```diff
@@ -88,13 +88,14 @@ public LedgerUnderreplicationManager newLedgerUnderreplicationManager()
     @Override
     public void format(AbstractConfiguration conf, LayoutManager layoutManager)
             throws InterruptedException, KeeperException, IOException {
-        String ledgersRootPath = conf.getZkLedgersRootPath();
-        List<String> children = zk.getChildren(ledgersRootPath, false);
-        for (String child : children) {
-            if (FlatLedgerManager.isSpecialZnode(child)) {
-                continue;
+        try (FlatLedgerManager ledgerManager = (FlatLedgerManager) newLedgerManager()) {
```
Comment (athanatos) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/FlatLedgerManagerFactory.java]: Why not make isLedgerParentNode static?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/HierarchicalLedgerManager.java [hunk 3]
```diff
@@ -86,6 +86,12 @@ protected long getLedgerId(String ledgerPath) throws IOException {
         return StringUtils.stringToLongHierarchicalLedgerId(hierarchicalPath);
     }
 
+    @Override
+    public boolean isSpecialZnode(String znode) {
```
Comment (athanatos) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/HierarchicalLedgerManager.java]: This would probably have to be legacyLM.isSpecialZnode(znode) || longLM.isSpecialZnode(znode) || super.isSpecialZnode(znode);

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookKeeperAdminTest.java [hunk 4]
```diff
@@ -230,4 +241,200 @@ public void testDecommissionForLedgersWithMultipleSegmentsAndNotWriteClosed() th
         }
         bkAdmin.close();
     }
+
+    @Test(timeout = 60000)
```
Comment (jiazhai) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookKeeperAdminTest.java]: Since we have a default timeout setting for all tests, it is suggested to remove `timeout` for individual test case.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LegacyHierarchicalLedgerManagerFactory.java [hunk 5]
```diff
@@ -79,7 +79,7 @@ public LedgerIdGenerator newLedgerIdGenerator() {
 
     @Override
     public LedgerManager newLedgerManager() {
-        return new HierarchicalLedgerManager(conf, zk);
```
Comment (jiazhai) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LegacyHierarchicalLedgerManagerFactory.java]: oh, the original code seems weird, why it return HierarchicalLedgerManager?

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookKeeperAdminTest.java [hunk 6]
```diff
@@ -230,4 +241,200 @@ public void testDecommissionForLedgersWithMultipleSegmentsAndNotWriteClosed() th
...
+        lh.close();
+
+        /*
+         * read instanceId
+         */
+        byte[] data = zkc.getData(newConfig.getZkLedgersRootPath() + "/" + BookKeeperConstants.INSTANCEID, false, null);
+        String readInstanceId = new String(data, UTF_8);
+
+        /*
+         * registrate a RO bookie
```
Comment (jiazhai) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookKeeperAdminTest.java]: typo? register

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java (hunk 1 above)
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java]: I didnt touch the implementation logic of this method, I just removed method arg for this method and try block.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/HierarchicalLedgerManager.java (hunk 3 above)
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/HierarchicalLedgerManager.java]: as we discussed, I'm merging all of these isSpecialZnode methods into one method, for the sake of simplicity and clarity.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookKeeperAdminTest.java (hunk 6 above)
Comment (reddycharan) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookKeeperAdminTest.java]: fixed it

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookKeeperAdminTest.java (hunk 4 above)
Comment (reddycharan) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookKeeperAdminTest.java]: what is the default time? and what if we want more time?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java [hunk 7]
```diff
@@ -421,94 +436,187 @@ public LayoutManager getLayoutManager(){
...
+        // create readonly bookies node if not exists
+        if (null == zk.exists(conf.getZkAvailableBookiesPath() + "/" + READONLY, false)) {
+            zk.create(conf.getZkAvailableBookiesPath() + "/" + READONLY, new byte[0], zkAcls, CreateMode.PERSISTENT);
+        }
+
+        return ledgerRootExists;
+    }
+
+    @Override
+    public boolean initNewCluster() throws Exception {
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java]: the logic here is problematic than the regular format. 

in regular format, it is crashing at the middle of a formatting op, you can re-run format to get it eventually done. in your implementation here, you can't. If I am crashing in the middle of a `init` op, I can't simply re-run `init`.

You might want to document this behavior on those two newly introduced command, to make sure people how to run the commands correctly.

Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java]: an alternative way to use `zk.txn` to make sure all the path creations are done in a multi op. but this probably is out of this scope, can be done in a separate task.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java [hunk 8]
```diff
@@ -421,94 +436,187 @@ public LayoutManager getLayoutManager(){
...
+                zkServers, zkLedgersRootPath);
+
+        boolean ledgerRootExists = null != zk.exists(conf.getZkLedgersRootPath(), false);
+        if (!ledgerRootExists) {
+            log.info("There is no existing cluster with ledgersRootPath: {} in ZKServers: {}, "
+                    + "so exiting nuke operation", zkLedgersRootPath, conf.getZkServers());
+            return true;
+        }
+
+        String availableBookiesPath = conf.getZkAvailableBookiesPath();
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java]: The logic here is way too complicated. You don't need to read the list the znodes and check if they are still exists. You can just issue a zk txn (don't remember that exactly method, but the following code is trying to demo how it works), if there are still bookies alive, the txn will fail.

```
Multi multi = zk.multi();
multi.delete(zkLedgersRootPath + "/available/readonly")
multi.delete(zkLedgersRootPath + "/available");

multi.execute();
```

Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java]: if you do as what I suggested, you don't need to change BookieWatcher.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookKeeperAdminTest.java (hunk 4 above)
Comment (sijie) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookKeeperAdminTest.java]: - default time is more than 60 seconds
- typically if a test is required more than 60 seconds, then we really need to think about how to improve the test. A lot of frakiness in current bookkeeper tests come from unnecessarily `Thread.sleep` or wait for something to trigger at a given time. we should try to avoid that pattern.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java [hunk 9]
```diff
@@ -59,10 +70,14 @@
...
 import org.apache.zookeeper.Watcher.Event.KeeperState;
 import org.apache.zookeeper.ZKUtil;
+import org.apache.zookeeper.ZooDefs.Ids;
 import org.apache.zookeeper.ZooKeeper;
 import org.apache.zookeeper.data.ACL;
 import org.apache.zookeeper.data.Stat;
 
+
+
+
```
Comment (athanatos) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java]: Stray whitespace?

Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java]: fixed it

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java (hunk 1 above)
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java]: for this commit, I would leave as it is

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LegacyHierarchicalLedgerManagerFactory.java (hunk 5 above)
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LegacyHierarchicalLedgerManagerFactory.java]: yes, I do see few other issues with 057af8dbce6c08794eb8b46ca52ca13f222d9bbb . Tried to address in this commit and we noticed other issues as well - Issue #1014

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookKeeperAdminTest.java (hunk 4 above)
Comment (reddycharan) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookKeeperAdminTest.java]: removed timeouts

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java (hunk 8 above)
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java]: Here before proceeding with nuking cluster data, I would like to make sure that there are no connected bookies to the cluster and print informative log message for each step/issue. Also, I moved deletions of the ZK nodes to  LedgerManagerFactory class, for the sake better abstraction. Anyhow the changes I made to BookieWatcher are minimal. For now, I feel comfortable to make independent zk operations, because check/delete operations are made at multiple classes with proper validation.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java (hunk 7 above)
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java]: I added enough logging when these operations fail, for the admin users to know why exactly it failed. Anyhow I'll add little more information to the issue description, regarding usage of these commands.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/LedgerManagerIteratorTest.java [hunk 10]
```diff
@@ -517,4 +557,20 @@ public void testLedgerManagerFormat() throws Throwable {
         Assert.assertTrue("ChildrenOfLedgersRootPathAfterFormat should contain all the invalid znodes created",
                 childrenOfLedgersRootPathAfterFormat.containsAll(invalidZnodes));
     }
+
+    @Test
+    public void hierarchicalLedgerManagerAsyncProcessLedgersTest() throws Throwable {
```
Comment (sijie) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/LedgerManagerIteratorTest.java]: nit: better to follow the naming as other methods here, so it is consistent.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/AbstractHierarchicalLedgerManager.java [hunk 11]
```diff
@@ -63,7 +63,11 @@ void asyncProcessLevelNodes(
         zk.sync(path, new AsyncCallback.VoidCallback() {
             @Override
             public void processResult(int rc, String path, Object ctx) {
-                if (rc != Code.OK.intValue()) {
+                if (rc == Code.NONODE.intValue()) {
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/AbstractHierarchicalLedgerManager.java]: These commits here seem not related to "decouple metaformat cmd". I would suggest they should go to a separate issue.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java (hunk 8 above)
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java]: > Here before proceeding with nuking cluster data, I would like to make sure that there are no connected bookies to the cluster and print informative log message for each step/issue.

the operation that I proposed is doing exactly same purpose as you hope, with much simpler approach and less error prone.

if you really feel the need to read and delete, you should not use BookieWatch, you should just use registration client api, rather than api in bookie watch. because bookie watch is the user of metadata interfaces, the metadata interface implementation should not use those api, this is a cycle dependency. 

These are the apis you should use: 

https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/RegistrationClient.java#L75
https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/RegistrationClient.java#L82

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java (hunk 7 above)
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java]: yes. we need enough instructions for people to run the command when they fail at the middle of running this command.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/AbstractHierarchicalLedgerManager.java (hunk 11 above)
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/AbstractHierarchicalLedgerManager.java]: ok..would move the 3rd commit to a new PR. But Bug1 in Issue #1014, would be fixed in this PR, since it is getting fixed as part of my changes for this PR. But Bug2 in Issue #1014 and testcases for both the bugs will be in new PR (which is currently 3rd commit).

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java (hunk 8 above)
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java]: reverting BookieWatcher changes and going to use RegistrationClient API

**Comment by reddycharan** [comment]
@sijie 

- I removed commit related to Issue #1014 from this PR
- reverted changes made to BookieWatcher, used RegistrationClient instead for getting RW and RO bookies in ZKRegistrationManager
- Will add new commands description/usage to the issue

**Comment by reddycharan** [comment]
@sijie you want me to provide more details to the issue description - https://github.com/apache/bookkeeper/issues/978 ?

**Comment by sijie** [comment]
@reddycharan no. I mean adding more descriptions in the commands. For example, if I failed when running initNewCluster, what I should do next? For users who don’t write the code, they will never not how to do. Because it is unlike meta format , you can run many times. You need to make instructions to ask people to nuke the cluster and reinit the cluster. Either you add the message at help message, or print this when failure happened. But we need something in the tool to tell users.

**Comment by reddycharan** [comment]
done adding suggested description and comments to the command. I think it should be good to go now.

**Comment by sijie** [comment]
@reddycharan can you rebase your branch? it should be ready to go.

**Comment by reddycharan** [comment]
rebased and checks are looking good

**Comment by reddycharan** [comment]
@sijie can you push this PR. Since this change touches quite a few files, I might run into rebasing issue again if it is delayed. Also, I've a bug fix which goes on top of this PR and I'm waiting for this PR to be closed.

I think other PRs of mine are also good to go.

**Comment by sijie** [comment]
@reddycharan sure. I would do it now. sorry I was busy yesterday, so didn't get any chances to go through those PRs.
